File: Java/103-BinaryTreeZigzagLevelOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        traverse(root, res, 0);
        return res;
    }
    
    void traverse(TreeNode curr, List<List<Integer>> res, int level){
        if (curr == null)   return;
        if (res.size() <= level){
            List<Integer> newLevel = new ArrayList<>();
            res.add(newLevel);
        }
        
        List<Integer> collection = res.get(level);
        if (level % 2 == 0) collection.add(curr.val);
        else    collection.add(0, curr.val);
        
        traverse(curr.left, res, level+1);
        traverse(curr.right, res, level+1);
    }
}

//Iterative
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null){
            return res;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int currLevel = 0;
        while (!queue.isEmpty()){
            int size = queue.size();
            List<Integer> tmp = new ArrayList<>();
            for (int i = 0; i < size; i++){
                TreeNode curr = queue.poll();
                if (currLevel % 2 == 0){
                    tmp.add(curr.val);
                }else{
                    tmp.add(0, curr.val);
                }
                if (curr.left != null){
                    queue.offer(curr.left);
                }
                if (curr.right != null){
                    queue.offer(curr.right);
                }
            }
            currLevel++;
            res.add(new ArrayList<>(tmp));
        }
        return res;
    }
}

File: cpp/103-BinaryTreeZigzagLevelOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (!root){
            return res;
        }
        queue<TreeNode*> queue;
        queue.push(root);
        int level = 0;
        
        while (!queue.empty()){
            vector<int> currLevel;
            int size = queue.size();
            for (int i = 0; i < size; i++){
                TreeNode* curr = queue.front();
                queue.pop();
                currLevel.push_back(curr->val);
                if (curr->left){
                    queue.push(curr->left);
                }
                if (curr->right){
                    queue.push(curr->right);
                }
            }
            if (level % 2 != 0){
                reverse(currLevel.begin(), currLevel.end());
            }
            res.push_back(currLevel);
            level++;
        }
        return res;
    }
};

//recursion
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (!root){
            return res;
        }
        traverse(root, res, 0);
        return res;
    }
    
    void traverse(TreeNode* root, vector<vector<int>> &res, int level){
        if (!root){
            return;
        }
        
        if (level >= res.size()){
            vector<int> tmp;
            res.push_back(tmp);
        }
        
        if (level % 2 == 0){
            res[level].push_back(root->val);
        }else{
            res[level].insert(res[level].begin(), root->val);
        }
        
        traverse(root->left, res, level+1);
        traverse(root->right, res, level+1);
    }
};